File: frontend/src/api/audit.ts
import { api } from './client'

export interface AuditEntry {
  id: number
  actor_id: string | null
  action: string
  resource_type: string
  resource_id: string | null
  payload: Record<string, unknown>
  prev_hash: string
  hash: string
  created_at: string
}

export interface AuditPage {
  items: AuditEntry[]
  total: number
  limit: number
  offset: number
}

export interface ChainStatus {
  ok: boolean
  broken_at: number | null
  total_entries: number
}

export interface AuditStats {
  total: number
  by_action: Record<string, number>
  by_resource: Record<string, number>
  first_ts: string | null
  last_ts: string | null
}

export interface AuditFilters {
  limit?: number
  offset?: number
  action?: string
  action_prefix?: string
  actor_id?: string
  resource_type?: string
  since?: string
  until?: string
}

function qs(params: AuditFilters): string {
  const sp = new URLSearchParams()
  Object.entries(params).forEach(([k, v]) => {
    if (v !== undefined && v !== '' && v !== null) sp.append(k, String(v))
  })
  const s = sp.toString()
  return s ? `?${s}` : ''
}

export const auditApi = {
  logs:   (f: AuditFilters = {}) => api.get<AuditPage>(`/audit/logs${qs(f)}`),
  verify: () => api.get<ChainStatus>('/audit/verify'),
  stats:  () => api.get<AuditStats>('/audit/stats'),
}

File: backend/app/api/audit.py
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from sqlalchemy import func, select

from app.api.deps import SessionDep, require_roles
from app.core.audit import verify_chain
from app.db.models.audit import AuditLog
from app.db.models.user import UserRole

router = APIRouter(prefix="/audit", tags=["audit"])

AdminOrAuditor = Annotated[
    object, Depends(require_roles(UserRole.admin, UserRole.auditor))
]


class AuditEntryOut(BaseModel):
    id: int
    actor_id: str | None
    action: str
    resource_type: str
    resource_id: str | None
    prev_hash: str
    hash: str
    created_at: str

    @classmethod
    def from_row(cls, r: AuditLog) -> "AuditEntryOut":
        return cls(
            id=r.id,
            actor_id=str(r.actor_id) if r.actor_id else None,
            action=r.action,
            resource_type=r.resource_type,
            resource_id=r.resource_id,
            prev_hash=r.prev_hash,
            hash=r.hash,
            created_at=r.created_at.isoformat(),
        )


class AuditPage(BaseModel):
    items: list[AuditEntryOut]
    total: int
    limit: int
    offset: int


class ChainStatus(BaseModel):
    ok: bool
    broken_at: int | None
    total_entries: int


@router.get("/logs", response_model=AuditPage)
async def list_logs(
    _: AdminOrAuditor,
    session: SessionDep,
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
) -> AuditPage:
    total = await session.scalar(select(func.count()).select_from(AuditLog)) or 0
    rows = (
        await session.scalars(
            select(AuditLog).order_by(AuditLog.id.desc()).limit(limit).offset(offset)
        )
    ).all()
    return AuditPage(
        items=[AuditEntryOut.from_row(r) for r in rows],
        total=int(total),
        limit=limit,
        offset=offset,
    )


@router.get("/verify", response_model=ChainStatus)
async def verify(_: AdminOrAuditor, session: SessionDep) -> ChainStatus:
    rows = (await session.scalars(select(AuditLog).order_by(AuditLog.id.asc()))).all()
    ok, broken = verify_chain(list(rows))
    return ChainStatus(ok=ok, broken_at=broken, total_entries=len(rows))
